datafusion/physical-expr/src/expressions/like.rs

```diff
@@ -0,0 +1,365 @@
+        context.with_boundaries(None)
+    }
+}
+
+impl PartialEq<dyn Any> for LikeExpr {
+    fn eq(&self, other: &dyn Any) -> bool {
+        down_cast_any_ref(other)
+            .downcast_ref::<Self>()
+            .map(|x| {
+                self.negated == x.negated
+                    && self.case_insensitive == x.case_insensitive
+                    && self.expr.eq(&x.expr)
+                    && self.pattern.eq(&x.pattern)
+            })
+            .unwrap_or(false)
+    }
+}
+
+macro_rules! binary_string_array_op_scalar {
+    ($LEFT:expr, $RIGHT:expr, $OP:ident, $OP_TYPE:expr) => {{
+        let result: Result<Arc<dyn Array>> = match $LEFT.data_type() {
+            DataType::Utf8 => compute_utf8_op_scalar!($LEFT, $RIGHT, $OP, StringArray, $OP_TYPE),
+            other => Err(DataFusionError::Internal(format!(
+                "Data type {:?} not supported for scalar operation '{}' on string array",
+                other, stringify!($OP)
+            ))),
+        };
+        Some(result)
+    }};
+}
+
+impl LikeExpr {
+    /// Evaluate the expression of the input is an array and
+    /// pattern is literal - use scalar operations
+    fn evaluate_array_scalar(
+        &self,
+        array: &dyn Array,
+        scalar: &ScalarValue,
+    ) -> Result<Option<Result<ArrayRef>>> {
+        let scalar_result = match (self.negated, self.case_insensitive) {
+            (false, false) => binary_string_array_op_scalar!(
+                array,
+                scalar.clone(),
+                like,
+                &DataType::Boolean
+            ),
+            (true, false) => binary_string_array_op_scalar!(
+                array,
+                scalar.clone(),
+                nlike,
+                &DataType::Boolean
+            ),
+            (false, true) => binary_string_array_op_scalar!(
+                array,
+                scalar.clone(),
+                ilike,
+                &DataType::Boolean
+            ),
+            (true, true) => binary_string_array_op_scalar!(
+                array,
+                scalar.clone(),
+                nilike,
+                &DataType::Boolean
+            ),
+        };
+        Ok(scalar_result)
+    }
+
+    fn evaluate_array_array(
+        &self,
+        left: Arc<dyn Array>,
+        right: Arc<dyn Array>,
+    ) -> Result<ArrayRef> {
+        match (self.negated, self.case_insensitive) {
+            (false, false) => binary_string_array_op!(left, right, like),
+            (true, false) => binary_string_array_op!(left, right, nlike),
+            (false, true) => binary_string_array_op!(left, right, ilike),
+            (true, true) => binary_string_array_op!(left, right, nilike),
+        }
+    }
+}
+
+/// Create a like expression whose arguments are correctly coerced.
+/// This function errors if it is not possible to coerce the arguments
+/// to computational types supported by the operator.
```

Comment:
I don't think this function does any coercion

Maybe something like this would be most appropriate:

```suggestion
/// Create a like expression, erroring if the argument types are not compatible
```